utils/import_data.py

```diff
@@ -52,123 +81,110 @@ def import_data(fw, df, dry_run=False):
 
     success_counter = 0
 
-    for row in range(nrows):
+    group_name = group.id
+    project_name = project.label
+
+    ############################################################################
+    # STEP 1: Loop through subject/session combos and aggregate file matches  #
+    ############################################################################
+    # We will first loop through and find any and all matches for each subject/session combination.
+    # Group by subject/session combos, to minimize loading.
+    unique_subjects = df[ROI.SUBJECT_HDR].unique()
+    # We are assuming that the group/project we're running in is the one we want to upload to.
+    initial_matching = {}
+    # First we will descend into the first unique subject label
+    for subject_label in unique_subjects:
+        # There may be multiple subjects, but they will all have the same label
+        subjects = project.subjects.iter_find(f'label={subject_label}')
+
+        # Now we will generate a dataframe group for the unique sessions present in this subject(s)
+        session_df = df[df[ROI.SUBJECT_HDR] == subject_label]
+        session_groups = session_df.groupby([ROI.SESSION_HDR])
+
+        # Now we will loop through any subjects we found originally
+        for subject in subjects:
+
+            # We will loop through our unique session groups
+            for session_label, indexs in session_groups.groups.items():
+                # And search for sessions with that label on that subject.
+                sessions = subject.sessions.iter_find(f'label={session_label}')
+
+                # We may find multiple - it is possible for 2 subjects to have the same label, each with a session with
+                #  the same label.
+                # Loop through the sessions we find (hopefully only one)
+                for session in sessions:
+
+                    # With each session, we must now search for each specific file
+                    for index in indexs:
+                        series = session_df.loc[index]
+                        object_name = series.get(ROI.MAPPING_COLUMN)
+
+                        lookup_string = f"{group_name}/{project_name}/{subject_label}/{session_label}/{object_name}"
+
+                        # Get a list of all files attached to the acquisitions in the session identified
+                        # by the object location found above
+                        objects_for_processing = fu.get_session_files(fw, session)
```

Comment:
I'm thinking this one can go outside the `for index in indexs:` loop, but inside the `for session in sessions:` loop

Reply:
good catch